File: pkg/cache/memory.go
package cache

import (
	"container/list"
	"context"
	"sync"
	"sync/atomic"
	"time"
)

// entry represents a single cache entry.
type entry struct {
	key       string
	value     any
	expiresAt time.Time
}

func (e *entry) isExpired() bool {
	if e.expiresAt.IsZero() {
		return false
	}
	return time.Now().After(e.expiresAt)
}

// MemoryCache is an in-memory LRU cache with TTL support.
type MemoryCache struct {
	mu              sync.RWMutex
	cache           map[string]*list.Element
	lru             *list.List
	maxSize         int
	defaultTTL      time.Duration
	cleanupInterval time.Duration
	stopCleanup     chan struct{}
	hits            atomic.Int64
	misses          atomic.Int64
}

// MemoryCacheOption is a functional option for MemoryCache.
type MemoryCacheOption func(*MemoryCache)

// WithMaxSize sets the maximum number of entries.
func WithMaxSize(size int) MemoryCacheOption {
	return func(c *MemoryCache) {
		c.maxSize = size
	}
}

// WithDefaultTTL sets the default TTL for entries.
func WithDefaultTTL(ttl time.Duration) MemoryCacheOption {
	return func(c *MemoryCache) {
		c.defaultTTL = ttl
	}
}

// WithCleanupInterval sets the interval for expired entry cleanup.
func WithCleanupInterval(interval time.Duration) MemoryCacheOption {
	return func(c *MemoryCache) {
		c.cleanupInterval = interval
	}
}

// NewMemoryCache creates a new in-memory cache.
func NewMemoryCache(opts ...MemoryCacheOption) *MemoryCache {
	c := &MemoryCache{
		cache:           make(map[string]*list.Element),
		lru:             list.New(),
		maxSize:         10000,
		defaultTTL:      time.Hour,
		cleanupInterval: time.Minute,
		stopCleanup:     make(chan struct{}),
	}

	for _, opt := range opts {
		opt(c)
	}

	// Start background cleanup goroutine
	go c.cleanupLoop()

	return c
}

func (c *MemoryCache) cleanupLoop() {
	ticker := time.NewTicker(c.cleanupInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			c.cleanupExpired()
		case <-c.stopCleanup:
			return
		}
	}
}

func (c *MemoryCache) cleanupExpired() {
	c.mu.Lock()
	defer c.mu.Unlock()

	for elem := c.lru.Front(); elem != nil; {
		next := elem.Next()
		e := elem.Value.(*entry)
		if e.isExpired() {
			c.lru.Remove(elem)
			delete(c.cache, e.key)
		}
		elem = next
	}
}

func (c *MemoryCache) evictIfNeeded() {
	for c.lru.Len() >= c.maxSize {
		oldest := c.lru.Front()
		if oldest != nil {
			e := oldest.Value.(*entry)
			c.lru.Remove(oldest)
			delete(c.cache, e.key)
		}
	}
}

// Get retrieves a value from the cache.
func (c *MemoryCache) Get(_ context.Context, key string) (any, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	elem, ok := c.cache[key]
	if !ok {
		c.misses.Add(1)
		return nil, nil
	}

	e := elem.Value.(*entry)
	if e.isExpired() {
		c.lru.Remove(elem)
		delete(c.cache, key)
		c.misses.Add(1)
		return nil, nil
	}

	// Move to back (most recently used)
	c.lru.MoveToBack(elem)
	c.hits.Add(1)
	return e.value, nil
}

// Set stores a value in the cache.
func (c *MemoryCache) Set(_ context.Context, key string, value any, ttl time.Duration) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if ttl == 0 {
		ttl = c.defaultTTL
	}

	var expiresAt time.Time
	if ttl > 0 {
		expiresAt = time.Now().Add(ttl)
	}

	// Check if key already exists
	if elem, ok := c.cache[key]; ok {
		c.lru.MoveToBack(elem)
		e := elem.Value.(*entry)
		e.value = value
		e.expiresAt = expiresAt
		return nil
	}

	// Evict if at capacity
	c.evictIfNeeded()

	// Add new entry
	e := &entry{
		key:       key,
		value:     value,
		expiresAt: expiresAt,
	}
	elem := c.lru.PushBack(e)
	c.cache[key] = elem

	return nil
}

// Delete removes a value from the cache.
func (c *MemoryCache) Delete(_ context.Context, key string) (bool, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	elem, ok := c.cache[key]
	if !ok {
		return false, nil
	}

	c.lru.Remove(elem)
	delete(c.cache, key)
	return true, nil
}

// Exists checks if a key exists in the cache.
func (c *MemoryCache) Exists(_ context.Context, key string) (bool, error) {
	c.mu.RLock()
	defer c.mu.RUnlock()

	elem, ok := c.cache[key]
	if !ok {
		return false, nil
	}

	e := elem.Value.(*entry)
	if e.isExpired() {
		return false, nil
	}

	return true, nil
}

// Clear removes all entries from the cache.
func (c *MemoryCache) Clear(_ context.Context) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	c.cache = make(map[string]*list.Element)
	c.lru.Init()
	return nil
}

// Close stops the cleanup goroutine and clears the cache.
func (c *MemoryCache) Close() error {
	close(c.stopCleanup)
	return c.Clear(context.Background())
}

// Size returns the current number of entries in the cache.
func (c *MemoryCache) Size() int {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.lru.Len()
}

// Stats returns cache statistics.
func (c *MemoryCache) Stats(_ context.Context) (Stats, error) {
	c.mu.RLock()
	defer c.mu.RUnlock()

	expiredCount := 0
	for elem := c.lru.Front(); elem != nil; elem = elem.Next() {
		e := elem.Value.(*entry)
		if e.isExpired() {
			expiredCount++
		}
	}

	return Stats{
		Size:         c.lru.Len(),
		MaxSize:      c.maxSize,
		ExpiredCount: expiredCount,
		Hits:         c.hits.Load(),
		Misses:       c.misses.Load(),
	}, nil
}

// GetMany retrieves multiple values from the cache.
func (c *MemoryCache) GetMany(ctx context.Context, keys []string) (map[string]any, error) {
	result := make(map[string]any, len(keys))
	for _, key := range keys {
		value, err := c.Get(ctx, key)
		if err != nil {
			return nil, err
		}
		if value != nil {
			result[key] = value
		}
	}
	return result, nil
}

// SetMany stores multiple values in the cache.
func (c *MemoryCache) SetMany(ctx context.Context, items map[string]any, ttl time.Duration) error {
	for key, value := range items {
		if err := c.Set(ctx, key, value, ttl); err != nil {
			return err
		}
	}
	return nil
}

// DeleteMany removes multiple values from the cache.
func (c *MemoryCache) DeleteMany(ctx context.Context, keys []string) (int, error) {
	count := 0
	for _, key := range keys {
		deleted, err := c.Delete(ctx, key)
		if err != nil {
			return count, err
		}
		if deleted {
			count++
		}
	}
	return count, nil
}

File: cpp/src/cache/memory.cpp
#include <retro_metadata/cache/memory.hpp>

namespace retro_metadata {

MemoryCache::MemoryCache(MemoryCacheOptions options) : options_(std::move(options)) {
    cleanup_thread_ = std::thread(&MemoryCache::cleanup_loop, this);
}

MemoryCache::~MemoryCache() {
    close();
}

MemoryCache::MemoryCache(MemoryCache&& other) noexcept
    : options_(std::move(other.options_)),
      cache_(std::move(other.cache_)),
      lru_(std::move(other.lru_)),
      hits_(other.hits_.load()),
      misses_(other.misses_.load()),
      stop_cleanup_(other.stop_cleanup_.load()) {
    other.stop_cleanup_ = true;
    if (other.cleanup_thread_.joinable()) {
        other.cleanup_thread_.join();
    }
    cleanup_thread_ = std::thread(&MemoryCache::cleanup_loop, this);
}

MemoryCache& MemoryCache::operator=(MemoryCache&& other) noexcept {
    if (this != &other) {
        close();

        options_ = std::move(other.options_);
        cache_ = std::move(other.cache_);
        lru_ = std::move(other.lru_);
        hits_ = other.hits_.load();
        misses_ = other.misses_.load();
        stop_cleanup_ = false;

        other.stop_cleanup_ = true;
        if (other.cleanup_thread_.joinable()) {
            other.cleanup_thread_.join();
        }
        cleanup_thread_ = std::thread(&MemoryCache::cleanup_loop, this);
    }
    return *this;
}

void MemoryCache::cleanup_loop() {
    while (!stop_cleanup_.load()) {
        std::this_thread::sleep_for(options_.cleanup_interval);
        if (!stop_cleanup_.load()) {
            cleanup_expired();
        }
    }
}

void MemoryCache::cleanup_expired() {
    std::unique_lock lock(mutex_);

    for (auto it = lru_.begin(); it != lru_.end();) {
        if (it->is_expired()) {
            cache_.erase(it->key);
            it = lru_.erase(it);
        } else {
            ++it;
        }
    }
}

void MemoryCache::evict_if_needed() {
    while (static_cast<int>(lru_.size()) >= options_.max_size) {
        if (lru_.empty()) break;
        auto oldest = lru_.begin();
        cache_.erase(oldest->key);
        lru_.erase(oldest);
    }
}

std::optional<std::any> MemoryCache::get(const std::string& key) {
    std::unique_lock lock(mutex_);

    auto it = cache_.find(key);
    if (it == cache_.end()) {
        misses_.fetch_add(1);
        return std::nullopt;
    }

    auto& entry = *it->second;
    if (entry.is_expired()) {
        lru_.erase(it->second);
        cache_.erase(it);
        misses_.fetch_add(1);
        return std::nullopt;
    }

    // Move to back (most recently used)
    lru_.splice(lru_.end(), lru_, it->second);
    hits_.fetch_add(1);
    return entry.value;
}

void MemoryCache::set(const std::string& key, std::any value, std::chrono::seconds ttl) {
    std::unique_lock lock(mutex_);

    if (ttl == std::chrono::seconds{0}) {
        ttl = options_.default_ttl;
    }

    std::chrono::steady_clock::time_point expires_at;
    if (ttl.count() > 0) {
        expires_at = std::chrono::steady_clock::now() + ttl;
    }

    // Check if key already exists
    auto it = cache_.find(key);
    if (it != cache_.end()) {
        lru_.splice(lru_.end(), lru_, it->second);
        it->second->value = std::move(value);
        it->second->expires_at = expires_at;
        return;
    }

    // Evict if at capacity
    evict_if_needed();

    // Add new entry
    lru_.push_back(Entry{key, std::move(value), expires_at});
    cache_[key] = std::prev(lru_.end());
}

bool MemoryCache::remove(const std::string& key) {
    std::unique_lock lock(mutex_);

    auto it = cache_.find(key);
    if (it == cache_.end()) {
        return false;
    }

    lru_.erase(it->second);
    cache_.erase(it);
    return true;
}

bool MemoryCache::exists(const std::string& key) {
    std::shared_lock lock(mutex_);

    auto it = cache_.find(key);
    if (it == cache_.end()) {
        return false;
    }

    return !it->second->is_expired();
}

void MemoryCache::clear() {
    std::unique_lock lock(mutex_);
    cache_.clear();
    lru_.clear();
}

void MemoryCache::close() {
    stop_cleanup_ = true;
    if (cleanup_thread_.joinable()) {
        cleanup_thread_.join();
    }
    clear();
}

CacheStats MemoryCache::stats() const {
    std::shared_lock lock(mutex_);

    int expired_count = 0;
    for (const auto& entry : lru_) {
        if (entry.is_expired()) {
            ++expired_count;
        }
    }

    return CacheStats{
        .size = static_cast<int>(lru_.size()),
        .max_size = options_.max_size,
        .expired_count = expired_count,
        .hits = hits_.load(),
        .misses = misses_.load()};
}

size_t MemoryCache::size() const {
    std::shared_lock lock(mutex_);
    return lru_.size();
}

std::unordered_map<std::string, std::any> MemoryCache::get_many(
    const std::vector<std::string>& keys) {
    std::unordered_map<std::string, std::any> result;
    for (const auto& key : keys) {
        auto value = get(key);
        if (value) {
            result[key] = std::move(*value);
        }
    }
    return result;
}

void MemoryCache::set_many(
    const std::unordered_map<std::string, std::any>& items, std::chrono::seconds ttl) {
    for (const auto& [key, value] : items) {
        set(key, value, ttl);
    }
}

int MemoryCache::delete_many(const std::vector<std::string>& keys) {
    int count = 0;
    for (const auto& key : keys) {
        if (remove(key)) {
            ++count;
        }
    }
    return count;
}

std::shared_ptr<MemoryCache> make_memory_cache() {
    return std::make_shared<MemoryCache>();
}

std::shared_ptr<MemoryCache> make_memory_cache(MemoryCacheOptions options) {
    return std::make_shared<MemoryCache>(std::move(options));
}

}  // namespace retro_metadata
